### ISSUE TITLE
Support for Oauth

### ISSUE THREAD

**Issue by polzka90** [fullstorydev/grpcui#99]
I am protecting my APIs with oauth2 token, is there any way to make the authentication process and storage the id_token to send as metadata to my APIs?

**Comment by jhump** [comment]
While oauth2 is a standard, it is also complicated and has multiple mechanisms for authorizing an application like grpcurl, so trying to build "one size fits all oauth2  support" in grpcurl is something I neither want to implement nor support.

I think the unix philosophy applies well here: instead of trying to teach grpcurl how to do oauth2 authn/authz and token generation, create a separate oauth2 command-line tool and compose them together. For example, if you had a tool that could issue access tokens, you could do something like so:

```
grpcurl -H "authorization: bearer $(oauth2 print-access-token)" some.server.com:443 some.Service/Method
```

This allows your oauth2 tool to do exactly what it needs to do and keeps grpcurl simple and minimal, while still providing the power of oauth2 support with grpcurl.

**Comment by venezia** [comment]
@jhump while I appreciate the desire to stay in one's lane, there is some precedent of similar tools (notably swagger-ui) that will facilitate an oauth2 encounter for the user.  On a few projects I work on we have a proto file using `grpc.gateway.protoc_gen_swagger.options.openapiv2_swagger` definitions and in conjunction with swagger-ui & grpc-gateway, we can have a developer log into swagger-ui and make api calls using the resulting JWT

While I don't think proto files have a native way to explain things like `grpc.gateway.protoc_gen_swagger.options.openapiv2_swagger.security_definition.security` does, it might be a great addition to grpc ui to offer something that approximates it, maybe we can even add an option to cover it too!.  Many APIs are protected by JWTs (as opposed to TLS, which seems to be the method grpc initially wanted individuals to handle authN)

What do you think about it?  Would you be open to a PR that adds this sort of functionality?

**Comment by jhump** [comment]
@venezia, I might be open to it, but it would be good to nail down the design a little further before writing a PR.

**Comment by augi** [comment]
We also run a bunch of gRPC services that use JWT for auth, namely we use Keycloak, and thus we would use the official Keycloak JavaScript client.

If I try to generalize these use-cases, it would be great to have the following _extension points_ in grpcui:
1. Allow specifying additional JavaScript files to be loaded to the page.
2. Provide a callback that would allow altering the outgoing request, e.g. add metadata.

@jhump , do these _extension points_ make sense for you? 🙏 Or is there even now a way how to do this? 🤔

**Comment by augi** [comment]
> ```
> grpcui -default-header "authorization: bearer $(oauth2 print-access-token)" some.server.com:443
> ```
This is actually what I'm using now but the issue is that the token has limited validity (for example 5 minutes). So I have to restart the server (that's OK) but the issue is that it obviously doesn't refresh my browser where the header is copied to  😄 

So it would be great if the _default headers_ wouldn't be copied to the Web form, and they would be appended automatically when the UI calls the `invoke` endpoint instead 🙏

**Comment by jhump** [comment]
@augi, instead of running something in the JS client, perhaps we could instead have a separate way for to grpcui to invoke arbitrary commands. For example, something like `-default-headers-command 'oauth2 print-access-token'`, where grpcui will then invoke the command `oauth2` with args `print-access-token` prior to each request and expect the output of the program to be headers, one per line.

I wish we could do something simpler, like just with unix pipes or something, so we don't end up having to implement a command interpreter in grpcui. To keep the implementation simple, the command would not support any special syntax/interpolation and perhaps not even allow whitespace in arguments (so no escaping necessary when parsing the args -- just split the string on whitespace to tokenize)... at least to start. If necessary, those improvements could be added later.

The main variance from `default-headers` is that this would actually inject headers into the request that are _never_ shown to the user in the web form -- basically secret/hidden headers. To that end, perhaps the argument name should be different, to avoid confusion. We'd also need a way to merge with user-supplied headers, in the event that the user entered a header with the same name into the web form. Would the merge logic simply combine (e.g. append all user-provided headers to the ones read from the program output), or would user-provided headers _override_ (e.g. remove headers read from the program output that collide with user-provided header names)?

WDYT?

**Comment by augi** [comment]
I actually see two use-cases, and I'm not sure if you want to support both, as the project maintainer.

1. The user executes the grpcui from their computer, so it acts like UI for grpcurl. We have this use case and we use a Docker container that is `FROM` grpcui:
```sh
#!/bin/sh
while true
do
  TOKEN=$(curl -s -X POST https://${KEYCLOAK_HOST:-our.keycloak.company.com}/auth/realms/${REALM:-master}/protocol/openid-connect/token -H "Content-Type: application/x-www-form-urlencoded" -d "username=${USERNAME}" -d "password=${PASSWORD}" -d 'grant_type=password' -d "client_id=${CLIENT_ID:-cli}" | jq -r '.access_token')
  timeout ${TIMEOUT:-300} /usr/bin/grpcui -bind=0.0.0.0 -port=${PORT:-8080} -default-header "Authorization: Bearer $TOKEN" "$@"
done
```
So it restarts the grpcui server every 5 minutes (expiration of the token). The obviously bad thing is that `-default-header` just prefills the value in the web form, so the server restart has no effect. What would help would be if there would be a `-hidden-default-header` argument that wouldn't be passed to the web form. Instead, it would be automatically appended by the grpcui server (probably in the `/invoke` handler), if not explicitly specified by user.

2. Second use-case is grpcui running somewhere (probably k8s in these days), and being used by many users. Then client-side authorization is required. And this is what this issue is about, IMHO. Maybe it could be sufficient to embed a generic OAuth2 client and allow to enable it using a cmd option.

**Comment by jhump** [comment]
Case 1 would the only one I'm particularly interested in maintaining since that is a use case that really needs a solution. Your proposed solution sounds very reasonable to me. Not only is it useful to have them inject server-side for your specific use case, but it's also useful to keep credentials secret (instead of including them in the web form), by having them only in-memory in the grpcui server. In fact, I think we could even name the argument `-secret-header` since that conveys that it's for metadata that is sensitive/secret.

For the second case, you could put an authenticating proxy in front of grpcui (something like https://github.com/oauth2-proxy/oauth2-proxy or https://github.com/kellegous/underpants). It looks like KeyCloak (referenced in your snippet) also [supports an authenticating proxy](https://eclipsesource.com/blogs/2018/01/11/authenticating-reverse-proxy-with-keycloak/). So for case 2, I don't see a compelling need for building that kind of logic into grpcui itself.

Having said that, a set of generic JS hooks plus the ability to inject custom assets (like JS files) could be a very useful "escape hatch" that allows people to do all kinds of crazy/clever things without having to add new features directly to grpcui. Maybe open a separate issue for that?

**Comment by augi** [comment]
Yes, the `-secret-header` sounds great! 👍 

The auth-proxy is a great idea, but as we want to pass the JWT token further to the called gRPC API, this would also require a JavaScript in grpcui, to _move_ the JWT token from a header to the form 🤔 
It would definitely work if we wanted just to protect the grpcui 👌

**Comment by jhump** [comment]
> this would also require a JavaScript in grpcui, to move the JWT token from a header to the form

Ah, I see. That makes sense. If those authenticating proxies don't include the token in a header to the backend, then indeed it must be client-side. If the JWT _does_ get passed from the proxy to the backend (which some proxies do, like Google's Identity-Aware-Proxy for Cloud Load Balancers), then it could be done server-side with another flag that tells grpcui what extra headers to propagate  -- in other words, headers that are in the "/invoke" HTTP request (but not _in_ the form body) that should also be added to outgoing metadata.

In any event, I think the general JS pluggability should be a separate issue and sounds like it would also let you solve the other angle of this.

**Comment by augi** [comment]
Yeah, you're right. In case of the proxy, it would be sufficient to _shovel_ the JWT in the `/invoke` handler (so no need for JS pluggability) 👍

**Comment by msuret** [comment]
I solved this by injecting Keycloak javascript client and the following code in grpcui web form using a browser plugin like [Code Injector](https://addons.mozilla.org/en-US/firefox/addon/codeinjector/):
```javascript
async function fillAuthorizationMetadata(){
    const keycloak = new Keycloak({
        url: 'https://my-kc-server.example.com/auth',
        realm: 'master',
        clientId: 'my-public-client'
    });
    const authenticated = await keycloak.init({ onLoad: 'login-required', checkLoginIframe: false })
    if(authenticated){
        document.getElementsByClassName("name")[1].value = "Authorization"
        document.getElementsByClassName("value")[1].value = `Bearer ${keycloak.token}`
        setInterval(async () => {
            await keycloak.updateToken(60);
            document.getElementsByClassName("value")[1].value = `Bearer ${keycloak.token}`
        }, 55000);
    }
}
fillAuthorizationMetadata();
```
Of course it would be much better if grpcui natively provided a way to use custom javascript files

**Comment by jhump** [comment]
While this issue is not directly solved, the features added in #163 and #164 should be enough to empower solutions _other than_ using browser extensions.
